### PR TITLE
[IE CLDNN] Activation with fused quantize bug fix

### PR DESCRIPTION
**Pull request by mzyczyns**
This change:
- fixes bug connected with quantization fusing to activation
- adds scale and activation fusing support
- adds corresponding tests

JIRA: 31506
Please review: @vparamuz @jhajducz @kdobros

### CODE REVIEW

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 1]
```diff
@@ -2931,37 +2930,66 @@ struct activation_test_params {
...
+#define CASE_ACTIVATION_I8_2 {1, 14, 8, 8}, data_types::i8, format::b_fs_yx_fsv16, activation_func::relu, {0.f, 0.f}, data_types::f32, format::bfyx
+#define CASE_ACTIVATION_3D_I8_1 {1, 17, 8, 8, 8}, data_types::i8, format::bfzyx, activation_func::relu, {0.f, 0.f}, data_types::f32, format::bfzyx
 
 class ActivationFusingTest : public ::BaseFusingTest<activation_test_params> {
 public:
     void execute(activation_test_params& p) {
         auto input_prim = get_mem(get_input_layout(p));
 
+        build_options options;
+        implementation_desc lrn_impl = {p.input_format, p.kernel_name};
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp:2988]: lrn_impl -> activation_impl (or some generic name)

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 2]
```diff
@@ -2989,94 +3032,233 @@ TEST_P(activation_quantize_i8, basic) {
...
+        activation_test_params{CASE_ACTIVATION_3D_F32_5, 2, 3, "activation_ref"},  // FIXME - accuracy bug
+        activation_test_params{CASE_ACTIVATION_3D_F32_6, 2, 3, "activation_ref"},  // FIXME - accuracy bug
+        activation_test_params{CASE_ACTIVATION_3D_F32_7, 2, 3, "activation_ref"},  // FIXME - accuracy bug
+        activation_test_params{CASE_ACTIVATION_3D_F32_8, 2, 3, "activation_ref"},  // FIXME - accuracy bug
     }), );
 
-class activation_opt_quantize_i8 : public ActivationFusingTest {};
-TEST_P(activation_opt_quantize_i8, basic) {
+class activation_quantize_scale_activation : public ActivationFusingTest {};
+TEST_P(activation_quantize_scale_activation, basic) {
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp:3086]: test name differs from the real topology used in the test

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 3]
```diff
@@ -2989,94 +3032,233 @@ TEST_P(activation_quantize_i8, basic) {
...
+        activation_test_params{CASE_ACTIVATION_3D_F32_2, 2, 5, "activation_ref"},
+        activation_test_params{CASE_ACTIVATION_3D_F32_3, 2, 5, "activation_ref"},
+        activation_test_params{CASE_ACTIVATION_3D_F32_4, 2, 5, "activation_ref"},
+    }), );
+
+INSTANTIATE_TEST_CASE_P(
+    DISABLED_fusings_gpu,
+    activation_quantize_scale_activation,
+    ::testing::ValuesIn(std::vector<activation_test_params>{
+        activation_test_params{CASE_ACTIVATION_F32_8, 2, 3, "activation_ref"},     // FIXME - accuracy bug
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp:3155]: `// FIXME - accuracy bug` is it some new accuracy bug?

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp [hunk 4]
```diff
@@ -179,8 +179,10 @@ void prepare_primitive_fusing::fuse_activations(program_impl &p) {
             // - primitives input cannot be output
             // - no activation additional input
             // - input was optimized
+            // - can't have fused primitives
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp:182]: Why? There's a branch that uses fused_ops mechanism if a node already has something fused

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 5]
```diff
@@ -2989,94 +3032,233 @@ TEST_P(activation_quantize_i8, basic) {
...
+        // InputDataType = FP32
+        activation_test_params{CASE_ACTIVATION_F32_0, 2, 3, "activation_opt"},
+        activation_test_params{CASE_ACTIVATION_F32_1, 2, 3, "activation_opt"},
+        activation_test_params{CASE_ACTIVATION_F32_2, 2, 3, "activation_opt"},
+        activation_test_params{CASE_ACTIVATION_F32_3, 2, 3, "activation_opt"},
+
+        activation_test_params{CASE_ACTIVATION_3D_F32_0, 2, 3, "activation_opt"},
+        activation_test_params{CASE_ACTIVATION_3D_F32_1, 2, 3, "activation_opt"},
+        activation_test_params{CASE_ACTIVATION_3D_F32_2, 2, 3, "activation_opt"},
+        activation_test_params{CASE_ACTIVATION_3D_F32_3, 2, 3, "activation_opt"},
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp:3054]: Is there a test case when activation_opt kernel is used and X % 4 != 0?

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp [hunk 6]
```diff
@@ -325,6 +327,40 @@ void prepare_primitive_fusing::fuse_simple_primitives(program_impl &p) {
             return false;
         };
 
+        auto activation_supports_fusings = [](activation_node& node) -> bool {
+            // We don't want to fuse activation when activation_kernel_opt is possible to use
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp:331]: Have you checked the performance? Is separate nodes with opt activation kernel really faster in cases like activation + quantize?

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 7]
```diff
@@ -2931,37 +2930,66 @@ struct activation_test_params {
...
+#define CASE_ACTIVATION_3D_F32_2 {1, 16, 8, 8, 8}, data_types::f32, format::bfzyx, activation_func::hard_sigmoid, {0.f, 0.f}, data_types::f32, format::bfzyx
+#define CASE_ACTIVATION_3D_F32_3 {1, 16, 8, 8, 8}, data_types::f32, format::bfzyx, activation_func::relu, {0.f, 0.f}, data_types::f32, format::bfzyx
+#define CASE_ACTIVATION_3D_F32_4 {1, 17, 7, 7, 7}, data_types::f32, format::b_fs_zyx_fsv16, activation_func::relu, {0.f, 0.f}, data_types::f32, format::bfzyx
+#define CASE_ACTIVATION_3D_F32_5 {1, 17, 7, 7, 7}, data_types::f32, format::b_fs_zyx_fsv32, activation_func::relu, {0.f, 0.f}, data_types::f32, format::bfzyx
+#define CASE_ACTIVATION_3D_F32_6 {1, 17, 7, 7, 7}, data_types::f32, format::bs_fs_yx_bsv16_fsv16, activation_func::relu, {0.f, 0.f}, data_types::f32, format::bfzyx
+#define CASE_ACTIVATION_3D_F32_7 {1, 17, 7, 7, 7}, data_types::f32, format::fs_b_yx_fsv32, activation_func::relu, {0.f, 0.f}, data_types::f32, format::bfzyx
+#define CASE_ACTIVATION_3D_F32_8 {1, 17, 7, 7, 7}, data_types::f32, format::fs_bs_yx_bsv4_fsv32, activation_func::relu, {0.f, 0.f}, data_types::f32, format::bfzyx
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp:2955]: Do we really need that much test cases for activation? The count of test cases here is pretty close to convolution's one...

E.g. CASE_ACTIVATION_3D_F32_{1,2,3} have exactly the same parameters except activation func, but different activation functions are supposed to be covered by activation primitive unit tests, not fusings tests.

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp (hunk 3 above)
Comment (mzyczyns) [reply inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp:3155]: Yes, it's connected with fusing formats in disabled tests to activation_ref kernel.

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp (hunk 4 above)
Comment (mzyczyns) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp:182]: I found that there is major accuracy bug when fusing activation to activation was added, which is beeing fixed by applying this change.

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp (hunk 6 above)
Comment (mzyczyns) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp:331]: **Performance results:**
  | Enabled fusing   execution time | Disabled fusing   execution time | Diff
-- | -- | -- | --
Sizes: | [us] | [us] | [us]
[b:1, f:128,   x:64, y:64, z:1, w:1, g:1] | 299.83 | 313.52 | -13.69
[b:1, f:512,   x:64, y:64, z:1, w:1, g:1] | 1294.18 | 1319.57 | -25.38
[b:1, f:1024,   x:64, y:64, z:1, w:1, g:1] | 2669.42 | 2730.30 | -60.88
[b:1, f:2048,   x:64, y:64, z:1, w:1, g:1] | 5505.50 | 5599.74 | -94.24
[b:1, f:4096,   x:64, y:64, z:1, w:1, g:1] | 11703.27 | 11757.30 | -54.03
[b:1, f:4096,   x:128, y:128, z:1, w:1, g:1] | 167022.00 | 166827.67 | 194.33
[b:1, f:4096,   x:256, y:256, z:1, w:1, g:1] | 783578.00 | 766798 | 16780.00
  |   |   |  


**Topology:**
- activation -> scale -> quantize(i8) -> reorder

When it comes to big size input data, disabling activation fusing support and using activaction_kernel_opt instead seems to make sense.

Whats your opinion?

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp (hunk 7 above)
Comment (mzyczyns) [reply inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp]: Test cases are now limited.

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp (hunk 6 above)
Comment (mzyczyns) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp]: Disabling fusing is removed.

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp (hunk 5 above)
Comment (mzyczyns) [reply inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp]: Added.

**Comment by vladimir-paramuzov** [comment]
@mzyczyns Could you rebase please?
I merged a fix for activation some time ago that disabled opt kernel when X*Y % 4 != 0 and !fused_ops.empty() (activation_kernel_opt.cpp:85)

I assume that this case is fixed by your patch, so probably this condition should be removed.

**Pull request merged**